### PR TITLE
warn when n(), row_number() or group_indices() are used without dplyr::

### PR DESCRIPTION
**Pull request by romainfrancois**
This is still the main problem that happens on the rev dev checks. This gives a warning rather than an error. 

``` r
dplyr::summarise(iris, n = n())
#> Warning: Calling n() without importing or prefixing it is deprecated
#>     n
#> 1 150
```

This would leave more time for dependent packages to adjust. 

Only doing it for `n()`, `row_number()` and `group_indices()` as this is I believe the only ones that are problematic on the rev dep checks, but perhaps this could be all dplyr functions that are possibly handled hybridly. That perhaps would be a little bit more code.

### CODE REVIEW

**Comment by romainfrancois** [comment]
This fixes some more 

```
> revdepcheck::revdep_check(num_workers=4)
── CHECK ──────────────────────────────────────────────────────── 63 packages ──
✖ carpenter 0.2.1                        ── E: 0  +1 | W: 0     | N: 0
✔ cleanNLP 2.3.0                         ── E: 0     | W: 0     | N: 0
✔ CollapseLevels 0.2.0                   ── E: 0     | W: 0     | N: 0
✖ compareDF 1.7.0                        ── E: 0  +1 | W: 0     | N: 1
✔ dat 0.4.0                              ── E: 0     | W: 0     | N: 0
✖ cytominer 0.1.0                        ── E: 0  +1 | W: 0     | N: 0
✖ dbplyr 1.3.0                           ── E: 0  +1 | W: 0     | N: 0-1
✖ desctable 0.1.3                        ── E: 0     | W: 0  +1 | N: 0
✔ ddpcr 1.11                             ── E: 0     | W: 0     | N: 1
✖ dexter 0.8.4                           ── E: 0     | W: 0  +1 | N: 0
✖ docxtools 0.2.0                        ── E: 0  +2 | W: 0     | N: 0
✖ dlookr 0.3.2                           ── E: 0  +1 | W: 0  +1 | N: 2
✖ dplyr.teradata 0.3.0                   ── E: 0     | W: 1  +1 | N: 1
✖ dtplyr 0.0.2                           ── E: 0  +1 | W: 0     | N: 0
✖ edeaR 0.8.1                            ── E: 0     | W: 0  +1 | N: 0
✔ episheet 0.4.0                         ── E: 0     | W: 0     | N: 0
✔ egor 0.18.08-02                        ── E: 1     | W: 1     | N: 0
✖ extdplyr 0.1.4                         ── E: 0  +1 | W: 0     | N: 0
✖ ezsummary 0.2.1                        ── E: 0  +2 | W: 0  +1 | N: 0
✖ evaluator 0.3.1                        ── E: 0  +1 | W: 0     | N: 0
✔ gaiah 0.0.2                            ── E: 0     | W: 0     | N: 1
✔ geomnet 0.2.0                          ── E: 0     | W: 0     | N: 0
✔ GerminaR 1.2                           ── E: 0     | W: 0     | N: 1
✔ GetITRData 0.7                         ── E: 0     | W: 0     | N: 0
✖ ggfan 0.1.2                            ── E: 0  +1 | W: 0     | N: 1
✖ GENESIS 2.10.1                         ── E: 1  +1 | W: 0     | N: 2
✔ grattan 1.7.0.0                        ── E: 0     | W: 0     | N: 1
✖ heemod 0.9.2                           ── E: 0  +2 | W: 0  +1 | N: 1
✖ IncucyteDRC 0.5.4                      ── E: 0  +1 | W: 0     | N: 0
✖ lmeresampler 0.1.0                     ── E: 0  +1 | W: 0     | N: 1
✔ MIAmaxent 1.0.0                        ── E: 0     | W: 0     | N: 1
✖ INDperform 0.1.1                       ── E: 0  +1 | W: 0     | N: 1
✖ modeldb 0.1.1                          ── E: 0     | W: 0  +1 | N: 0
✔ mousetrap 3.1.1                        ── E: 0     | W: 0     | N: 0
✖ naniar 0.4.1                           ── E: 0  +2 | W: 0     | N: 0
✖ pixiedust 0.8.6                        ── E: 0  +1 | W: 0     | N: 0
✖ poplite 0.99.19                        ── E: 0  +1 | W: 0  +1 | N: 0
✖ processcheckR 0.1.0                    ── E: 0     | W: 0  +1 | N: 0
✖ portalr 0.2.1                          ── E: 0  +1 | W: 0     | N: 0
✖ purrrlyr 0.0.3                         ── E: 0  +2 | W: 0     | N: 0
✖ psychmeta 2.3.0                        ── E: 0  +1 | W: 0     | N: 1
✖ radiant.model 0.9.7                    ── E: 0  +2 | W: 0     | N: 0
✔ rbin 0.1.1                             ── E: 0     | W: 0     | N: 1
✔ rcv 0.2.1                              ── E: 0     | W: 0     | N: 2
✖ replyr 0.9.9                           ── E: 0  +1 | W: 1     | N: 1
✖ rPref 1.2                              ── E: 0  +1 | W: 0  +1 | N: 1
✖ ruler 0.1.4                            ── E: 0  +1 | W: 0  +1 | N: 0
✖ RWDataPlyr 0.6.2                       ── E: 0  +1 | W: 0     | N: 0
✔ rubias 0.1.0                           ── E: 0     | W: 0     | N: 1
✔ seqcombo 1.2.0                         ── E: 0     | W: 0     | N: 0
✔ seqCAT 1.2.1                           ── E: 0     | W: 1     | N: 0
✖ SeqVarTools 1.18.0                     ── E: 0  +2 | W: 1     | N: 1
✖ spdplyr 0.1.3                          ── E: 0  +1 | W: 0     | N: 0
✔ sjPlot 2.6.2                           ── E: 0     | W: 0     | N: 0
✖ tbrf 0.1.1                             ── E: 0  +1 | W: 0     | N: 0
✖ tibbletime 0.1.1                       ── E: 0  +1 | W: 0     | N: 0
✖ tidygraph 1.1.1                        ── E: 0     | W: 0  +1 | N: 0
✖ treeplyr 0.1.4                         ── E: 0  +1 | W: 0     | N: 0
✔ tidytransit 0.3.5                      ── E: 0     | W: 0     | N: 3
✖ tsibble 0.6.2                          ── E: 0  +1 | W: 0     | N: 0
✖ visdat 0.5.2                           ── E: 0  +2 | W: 0     | N: 1
✖ vqtl 2.0.4                             ── E: 0  +1 | W: 0     | N: 2
✖ zFactor 0.1.7                          ── E: 0  +2 | W: 0     | N: 1
OK: 1389
BROKEN: 102
Total time: 49 min
```

**Comment by romainfrancois** [comment]
This however only catches cases when e.g. `n()` is called directly, hence this error on `carpenter`: 

```
> revdepcheck::revdep_details(revdep = "carpenter")
══ Reverse dependency check ════════════════════════════════════════════════════════════════════ carpenter 0.2.1 ══

Status: BROKEN

── Newly failing

✖ checking tests ...

── Before ─────────────────────────────────────────────────────────────────────────────────────────────────────────
0 errors ✔ | 0 warnings ✔ | 0 notes ✔

── After ──────────────────────────────────────────────────────────────────────────────────────────────────────────
❯ checking tests ...
  See below...

── Test failures ──────────────────────────────────────────────────────────────────────────────────── testthat ────

> library(testthat)
> library(carpenter)
> 
> test_check("carpenter")
── 1. Error: build_table outputs correct information (@test-build_tables.R#21)  
could not find function "n"
1: build_table(draft, finish = FALSE) at testthat/test-build_tables.R:21
2: dplyr::bind_rows(lapply(names(sketch$rows), function(x, s = sketch) {
       func <- eval(parse(text = x))
       sketch_rows <- s$rows[[x]]
       type <- sketch_rows$type
       switch(type, numeric = {
           row_create <- make_numeric_row
       }, factor = {
           row_create <- make_factor_row
       })
       row_create(data = data, header = s$header, rows = sketch_rows$vars, stat = func, 
           digits = sketch_rows$digits)
   })) at /Users/romain/git/tidyverse/dplyr-revdep/revdep/checks.noindex/carpenter/new/carpenter.Rcheck/00_pkg_src/carpenter/R/build_table.R:31
3: flatten_bindable(dots_values(...)) at /Users/romain/git/tidyverse/dplyr/R/bind.r:99
4: dots_values(...) at /Users/romain/git/tidyverse/dplyr/R/bind.r:99
5: lapply(names(sketch$rows), function(x, s = sketch) {
       func <- eval(parse(text = x))
       sketch_rows <- s$rows[[x]]
       type <- sketch_rows$type
       switch(type, numeric = {
           row_create <- make_numeric_row
       }, factor = {
           row_create <- make_factor_row
       })
       row_create(data = data, header = s$header, rows = sketch_rows$vars, stat = func, 
           digits = sketch_rows$digits)
   })
6: FUN(X[[i]], ...)
7: row_create(data = data, header = s$header, rows = sketch_rows$vars, stat = func, 
       digits = sketch_rows$digits) at /Users/romain/git/tidyverse/dplyr-revdep/revdep/checks.noindex/carpenter/new/carpenter.Rcheck/00_pkg_src/carpenter/R/build_table.R:43
8: dplyr::group_by_(data, header, "Variables") %>% dplyr::mutate_(n = lazyeval::interp("stat(n, digits)", 
       stat = stat, digits = digits)) %>% tidyr::spread_(header, "n") %>% dplyr::ungroup() %>% 
       dplyr::mutate_(id = lazyeval::interp("1:n()")) at /Users/romain/git/tidyverse/dplyr-revdep/revdep/checks.noindex/carpenter/new/carpenter.Rcheck/00_pkg_src/carpenter/R/build_table.R:109
9: withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
10: eval(quote(`_fseq`(`_lhs`)), env, env)
11: eval(quote(`_fseq`(`_lhs`)), env, env)
12: `_fseq`(`_lhs`)
13: freduce(value, `_function_list`)
14: withVisible(function_list[[k]](value))
15: function_list[[k]](value)
16: dplyr::mutate_(., id = lazyeval::interp("1:n()"))
17: mutate_.tbl_df(., id = lazyeval::interp("1:n()")) at /Users/romain/git/tidyverse/dplyr/R/manip.r:457
18: mutate_impl(.data, dots) at /Users/romain/git/tidyverse/dplyr/R/tbl-df.r:96

══ testthat results  ═══════════════════════════════════════════════════════════
OK: 19 SKIPPED: 0 FAILED: 1
1. Error: build_table outputs correct information (@test-build_tables.R#21) 

Error: testthat unit tests failed
Execution halted

1 error ✖ | 0 warnings ✔ | 0 notes ✔
```

File: inst/include/dplyr/hybrid/Expression.h [hunk 1]
```diff
@@ -110,18 +110,40 @@ class Expression {
...
-        dplyr_hash_map<SEXP, hybrid_function>::const_iterator it = get_hybrid_inline_map().find(f);
-        if (it != get_hybrid_inline_map().end()) {
-          func = it->second.name;
-          package = it->second.package;
-          id = it->second.id;
+          if (workaround) {
+            func = head;
+            package = symbols::dplyr;
+            id = get_hybrid_named_map().find(func)->second.id;
+          }
```
Comment (krlmlr) [inline inst/include/dplyr/hybrid/Expression.h]: What happens if `!workaround` ?

File: inst/include/dplyr/hybrid/Expression.h [hunk 2]
```diff
@@ -110,18 +110,40 @@ class Expression {
       // then match that against the hash map
       FindFunData finder(head, env);
       if (finder.findFun()) {
+        if (Rf_isNull(finder.res)) {
+          // no function was found, but
+          //
+          // handle n(), row_number(), group_indices() in case dplyr is not imported
+          bool workaround = true;
+          if (head == symbols::n) {
+            Rcpp::warningcall(R_NilValue, "Calling n() without importing or prefixing it is deprecated");
```
Comment (krlmlr) [inline inst/include/dplyr/hybrid/Expression.h]:
```suggestion
            Rcpp::warningcall(R_NilValue, "Calling `n()` without importing or prefixing it is deprecated, use `dplyr::n()`.");
```

File: inst/include/dplyr/hybrid/Expression.h (hunk 1 above)
Comment (romainfrancois) [reply inst/include/dplyr/hybrid/Expression.h]: Nothing. It just ends up going through standard r evaluation.

**Comment by hadley** [comment]
Can you please double check with @lionel- that this is the same approach we use in rlang so developers will still get warnings when testing their own packages locally?

**Comment by romainfrancois** [comment]
probably not. Here the warnings are unconditional, I suppose this could call `warn_deprecated()` or `signal_soft_deprecated()` instead of `Rf_warningcall`

**Comment by lionel-** [comment]
Yes it'd need to call into R. We should eventually export C callables from the lifecycle package.

**Comment by romainfrancois** [comment]
Please have a look at #4153

**Comment by lionel-** [comment]
Note that with `warn_deprecated()`, the warnings are displayed unconditionally, only once per session. If you'd like to avoid the warnings on CRAN machines or across packages, you need `signal_soft_deprecated()`.

**Comment by romainfrancois** [comment]
I think we can go straight to `warn_deprecated()` for this.

**Comment by hadley** [comment]
Won't a warning on a CRAN machine re-create the original problem?

**Comment by romainfrancois** [comment]
I see. 

I can make this a `signal_soft_deprecated()` for 0.8.0 so that it makes no noise on cran now and gives a chance to maintainers to see and maybe fix it. 

I was keen on going directly to `warn_deprecated()` because this was already improvement over getting an error.

**Comment by romainfrancois** [comment]
#4156

**Pull request merged**